notebook2/notebook/notebook.py

```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
     return url + path
 
 
+# FIXME: use hash
+def UNSAFE_user_id_transform(user_id): return user_id.replace('|', '--_--')
```

Comment:
Issue: https://github.com/hail-is/hail/issues/5377

Reply:
I'm inclined to not leave a FIXME in since the method is already labeled UNSAFE and we have an issue.

Reply:
Got it, agreed

Reply:
Ok, so the reason I didn't do this yet...I'm a bit worried about the probability of collisions.

The reason I think this is unsafe is kind of for the same reason, that I will have collisions by replacing all "|" with some other pattern. 

Auth0 ids come in the form of provider|id. One solution is to replace only the first |. A better solution seems to encode into a form that kube will accept.

Do you have any recommendations?

Reply:
heh, I really hate when people don't precisely define the syntax of their data.
Maybe this?
```python
x = re.match(r"^([^|]+)\|(.+)$", user_id)
identity_provider = x.group(1)
user_id = x.group(2)
identity_provider + '.' + user_id
```

? Is there an email or forum we can ask about the syntax of identity_providers? I've only seen dashes and characters.

Reply:
Agreed, good suggestion. Did a bit of research.
1) Auth0's documentation states <provier>|<id> , but doesn't specify exactly what the id format is (as far as I can see): https://auth0.com/docs/tokens/id-token

2) [The OpenID Connect standard](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims), which Auth0 implements (although I'm not sure all of their APIs are compliant), specifies ASCII and 255 characters max, which would be disallowed by Kube's label restrictions, if full range is used.

3) [I opened an issue to ask](https://community.auth0.com/t/sub-claim-format-kubernetes-restricted-set/21996)

For now, I hashed using SHA224 and store the hex digest. What do you think of that?

Reply:
Heh. The response side-stepped defining what the identity provider's syntax is. 🤷‍♀️ 

SHA seems fine.

Reply:
Actually wait, why are we introducing a possibly colliding hash? We can just base64 encode the bytes, no?

```
In [3]: base64.b64decode(base64.b64encode("google-foogle|bobadob"))
Out[3]: 'google-foogle|bobadob'
```

We pay a 4/3x inflation, which leaves us about 189 characters for the tokens.

Reply:
Cotton and I had spoken about it: https://github.com/hail-is/hail/pull/5215#issuecomment-459225923 , it can be used with some restrictions. However, I think it may be more difficult than is outlined, because we need to take care of note only padding (`=`), but also `+` and `/`, which are in the output grammar (there some encodings that use `/` and `.`). Kubernetes allows `.`, `_`, and `-`, but requires those be in-between [a-Z0-9] characters, which I'm not sure we can guarantee.

In playing with character encodings, I can't actually get a + generated except in between alphanumeric ascii characters, but I would need to formally show that it isn't possible.